File: go/test/protocol/telemetry_sender_tester.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
package protocol

import (
	"context"
	"fmt"
	"log"
	"net/url"
	"os"
	"path/filepath"
	"slices"
	"strings"
	"testing"

	"github.com/Azure/iot-operations-sdks/go/protocol"
	"github.com/Azure/iot-operations-sdks/go/protocol/errors"
	"github.com/BurntSushi/toml"
	"github.com/relvacode/iso8601"
	"github.com/stretchr/testify/require"
	"gopkg.in/yaml.v3"
)

func RunTelemetrySenderTests(t *testing.T) {
	var telemetrySenderDefaultInfo DefaultTestCase

	_, err := toml.DecodeFile(
		"../../../eng/test/test-cases/Protocol/TelemetrySender/defaults.toml",
		&telemetrySenderDefaultInfo,
	)
	if err != nil {
		panic(err)
	}

	TestCaseDefaultInfo = &telemetrySenderDefaultInfo
	TestCaseDefaultSerializer = &telemetrySenderDefaultInfo.Prologue.Sender.Serializer

	files, err := filepath.Glob(
		"../../../eng/test/test-cases/Protocol/TelemetrySender/*.yaml",
	)
	if err != nil {
		log.Fatal(err)
	}

	for ix, f := range files {
		testName, _ := strings.CutSuffix(filepath.Base(f), ".yaml")
		t.Run(testName, func(t *testing.T) {
			runOneTelemetrySenderTest(t, ix, testName, f)
		})
	}
}

func runOneTelemetrySenderTest(
	t *testing.T,
	testCaseIndex int,
	testName string,
	fileName string,
) {
	pendingTestCases := []string{
		"TelemetrySenderPubAckDroppedByDisconnection_ReconnectAndSuccess", // hangs intermittently
		"TelemetrySenderPubAckFailure_ThrowsException",                    // perhaps related to https://github.com/eclipse/paho.golang/issues/216
	}

	testCaseYaml, err := os.ReadFile(fileName)
	if err != nil {
		log.Fatal(err)
	}

	//	t.Parallel()

	var testCase TestCase
	err = yaml.Unmarshal(testCaseYaml, &testCase)
	if err != nil {
		log.Fatal(err)
	}

	if slices.Contains(testCase.Requires, Unobtanium) ||
		slices.Contains(testCase.Requires, ExplicitDefault) {
		t.Skipf(
			"Skipping test %s because it requires an unavailable feature",
			testName,
		)
	}

	if slices.Contains(pendingTestCases, testName) {
		t.Skipf(
			"Skipping test %s because it requires a feature which has not yet been implemented",
			testName,
		)
	}

	var mqttClientID string
	if testCase.Prologue.MqttConfig.ClientID != nil {
		mqttClientID = *testCase.Prologue.MqttConfig.ClientID
	} else {
		mqttClientID = fmt.Sprintf("SenderTestClient%d", testCaseIndex)
	}

	stubBroker, sessionClient := getStubAndSessionClient(t, mqttClientID)

	for _, ackKind := range testCase.Prologue.PushAcks.Publish {
		stubBroker.EnqueuePubAck(ackKind)
	}

	for _, ackKind := range testCase.Prologue.PushAcks.Subscribe {
		stubBroker.EnqueueSubAck(ackKind)
	}

	for _, ackKind := range testCase.Prologue.PushAcks.Unsubscribe {
		stubBroker.EnqueueUnsubAck(ackKind)
	}

	telemetrySenders := make(map[string]*TestingTelemetrySender)

	for ix, tcs := range testCase.Prologue.Senders {
		var catch *TestCaseCatch
		if ix == len(testCase.Prologue.Senders)-1 {
			catch = testCase.Prologue.Catch
		}

		sender := getTelemetrySender(t, sessionClient, stubBroker, &tcs, catch)
		if sender != nil {
			telemetrySenders[*tcs.TelemetryName] = sender
		}
	}

	sendChan := make(chan error)

	for _, action := range testCase.Actions {
		switch action.Kind {
		case SendTelemetry:
			sendTelemetry(
				t,
				action.AsSendTelemetry(),
				telemetrySenders,
				sendChan,
			)
		case AwaitSend:
			awaitSend(t, action.AsAwaitSend(), sendChan)
		case AwaitPublish:
			awaitPublishTelemetry(
				t,
				action.AsAwaitPublish(),
				stubBroker,
				map[int][]byte{},
			)
		case Disconnect:
			stubBroker.Disconnect()
		}
	}

	close(sendChan)

	for err := range sendChan {
		require.NoErrorf(
			t,
			err,
			"Unexpected error returned when awaiting TelemetrySender.Send()",
		)
	}

	if testCase.Epilogue.PublicationCount != nil {
		require.Equal(
			t,
			*testCase.Epilogue.PublicationCount,
			stubBroker.PublicationCount,
		)
	}

	for ix, publishedMessage := range testCase.Epilogue.PublishedMessages {
		checkPublishedTelemetry(t, ix, &publishedMessage, stubBroker)
	}
}

func getTelemetrySender(
	t *testing.T,
	sessionClient protocol.MqttClient,
	stubBroker *StubBroker,
	tcs *TestCaseSender,
	catch *TestCaseCatch,
) *TestingTelemetrySender {
	options := []protocol.TelemetrySenderOption{
		protocol.WithTopicTokens(tcs.TopicTokenMap),
	}

	if tcs.TopicNamespace != nil {
		options = append(
			options,
			protocol.WithTopicNamespace(*tcs.TopicNamespace),
		)
	}

	sender, err := NewTestingTelemetrySender(
		sessionClient,
		&tcs.Serializer,
		tcs.TelemetryTopic,
		options...)

	if catch == nil {
		require.NoErrorf(
			t,
			err,
			"Unexpected error returned when initializing TelemetrySender: %s",
			err,
		)
	} else {
		if err == nil {
			err = sender.base.Send(context.Background(), *TestCaseDefaultInfo.Actions.SendTelemetry.GetTelemetryValue())
			stubBroker.AwaitPublish()
		}

		require.Errorf(t, err, "Expected %s error, but no error returned when initializing TelemetrySender", catch.ErrorKind)
		CheckError(t, *catch, err)
	}

	return sender
}

func sendTelemetry(
	t *testing.T,
	actionSendTelemetry *TestCaseActionSendTelemetry,
	telemetrySenders map[string]*TestingTelemetrySender,
	sendChan chan error,
) {
	telem := *actionSendTelemetry.TelemetryValue

	options := []protocol.SendOption{}
	options = append(
		options,
		protocol.WithTimeout(actionSendTelemetry.Timeout.ToDuration()),
		protocol.WithTopicTokens(actionSendTelemetry.TopicTokenMap),
	)

	if actionSendTelemetry.Qos != nil && *actionSendTelemetry.Qos != 1 {
		t.Skipf(
			"Skipping test because TelemetrySender does not support settable QoS",
		)
	}

	if actionSendTelemetry.Metadata != nil {
		for key, val := range *actionSendTelemetry.Metadata {
			options = append(
				options,
				protocol.WithMetadata{
					key: val,
				},
			)
		}
	}

	if actionSendTelemetry.CloudEvent != nil {
		cloudEvent := protocol.CloudEvent{}

		if actionSendTelemetry.CloudEvent.Source != nil {
			source, err := url.Parse(*actionSendTelemetry.CloudEvent.Source)
			if err != nil {
				go func() { sendChan <- getCloudEventError("Source", *actionSendTelemetry.CloudEvent.Source, "URL", err) }()
				return
			}
			cloudEvent.Source = source
		}

		if actionSendTelemetry.CloudEvent.SpecVersion != nil {
			cloudEvent.SpecVersion = *actionSendTelemetry.CloudEvent.SpecVersion
		}

		if actionSendTelemetry.CloudEvent.Type != nil {
			cloudEvent.Type = *actionSendTelemetry.CloudEvent.Type
		}

		if actionSendTelemetry.CloudEvent.ID != nil {
			cloudEvent.ID = *actionSendTelemetry.CloudEvent.ID
		}

		if time, ok := actionSendTelemetry.CloudEvent.Time.(string); ok {
			isoTime, err := iso8601.ParseString(time)
			if err != nil {
				go func() { sendChan <- getCloudEventError("Time", time, "ISO 8601", err) }()
				return
			}
			cloudEvent.Time = isoTime
		}

		if subject, ok := actionSendTelemetry.CloudEvent.Subject.(string); ok {
			cloudEvent.Subject = subject
		}

		if dataSchema, ok := actionSendTelemetry.CloudEvent.DataSchema.(string); ok {
			dataSchemaURL, err := url.Parse(dataSchema)
			if err != nil {
				go func() { sendChan <- getCloudEventError("DataSchema", dataSchema, "URL", err) }()
				return
			}
			cloudEvent.DataSchema = dataSchemaURL
		}

		options = append(
			options,
			protocol.WithCloudEvent(&cloudEvent),
		)
	}

	tcs := telemetrySenders[*actionSendTelemetry.TelemetryName]

	go func() {
		err := tcs.base.Send(
			context.Background(),
			telem,
			options...)
		sendChan <- err
	}()
}

func awaitSend(
	t *testing.T,
	actionAwaitSend *TestCaseActionAwaitSend,
	sendChan chan error,
) {
	err := <-sendChan

	if actionAwaitSend.Catch == nil {
		require.NoErrorf(
			t,
			err,
			"Unexpected error returned when awaiting TelemetrySender.Send()",
		)
	} else {
		require.Errorf(t, err, "Expected %s error, but no error returned when awaiting TelemetrySender.Send()", actionAwaitSend.Catch.ErrorKind)
		CheckError(t, *actionAwaitSend.Catch, err)
	}
}

func awaitPublishTelemetry(
	_ *testing.T,
	_ *TestCaseActionAwaitPublish,
	stubBroker *StubBroker,
	_ map[int][]byte,
) {
	stubBroker.AwaitPublish()
}

func checkPublishedTelemetry(
	t *testing.T,
	sequenceIndex int,
	publishedMessage *TestCasePublishedMessage,
	stubBroker *StubBroker,
) {
	msg, ok := stubBroker.GetIndexedPublishedMessage(sequenceIndex)
	require.True(t, ok)

	if publishedMessage.Topic != nil {
		require.Equal(t, *publishedMessage.Topic, msg.Topic)
	}

	if publishedMessage.Payload == nil {
		require.Empty(t, msg.Payload)
	} else if payload, ok := publishedMessage.Payload.(string); ok {
		require.Equal(t, payload, string(msg.Payload))
	}

	if publishedMessage.ContentType != nil {
		require.Equal(
			t,
			*publishedMessage.ContentType,
			msg.Properties.ContentType,
		)
	}

	if publishedMessage.FormatIndicator != nil {
		require.Equal(
			t,
			*publishedMessage.FormatIndicator,
			*msg.Properties.PayloadFormat,
		)
	}

	for key, val := range publishedMessage.Metadata {
		propVal, ok := getUserProperty(t, msg, key)
		if val != nil {
			require.True(t, ok)
			require.Equal(t, *val, propVal)
		} else {
			require.False(t, ok)
		}
	}

	if publishedMessage.SourceID != nil {
		sourceID, ok := getUserProperty(t, msg, SourceID)
		require.True(t, ok)
		require.Equal(t, *publishedMessage.SourceID, sourceID)
	}

	if publishedMessage.Expiry != nil {
		require.Equal(
			t,
			*publishedMessage.Expiry,
			*msg.Properties.MessageExpiry,
		)
	}
}

func getCloudEventError(
	fieldName string,
	propValue string,
	parseType string,
	err error,
) error {
	return &errors.Client{
		Message: fmt.Sprintf(
			"cloud event %s not parsable as %s",
			fieldName,
			parseType,
		),
		Kind: errors.ConfigurationInvalid{
			PropertyName:  "CloudEvent",
			PropertyValue: propValue,
		},
		Nested:  err,
		Shallow: true,
	}
}

File: dotnet/test/Azure.Iot.Operations.Protocol.MetlTests/TelemetrySenderTester.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System.Globalization;
using System.Text;
using System.Threading.Tasks;
using Azure.Iot.Operations.Mqtt.Converters;
using Azure.Iot.Operations.Protocol.Models;
using Azure.Iot.Operations.Protocol.Telemetry;
using Microsoft.VisualStudio.Threading;
using Tomlyn;
using Xunit;
using YamlDotNet.Core;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

namespace Azure.Iot.Operations.Protocol.MetlTests
{
    public class TelemetrySenderTester
    {
        private const string testCasesPath = "../../../../../../eng/test/test-cases";
        private const string senderCasesPath = $"{testCasesPath}/Protocol/TelemetrySender";
        private const string defaultsFilePath = $"{testCasesPath}/Protocol/TelemetrySender/defaults.toml";

        private static readonly TimeSpan TestTimeout = TimeSpan.FromMinutes(1);

        private static readonly HashSet<string> problematicTestCases = new HashSet<string>
        {
            "TelemetrySenderSendWithCloudEventDataContentTypeNonConforming_ThrowsException",
            "TelemetrySenderSendWithCloudEventDataSchemaEmpty_ThrowsException",
            "TelemetrySenderSendWithCloudEventDataSchemaNonUri_ThrowsException",
            "TelemetrySenderSendWithCloudEventDataSchemaRelativeUri_ThrowsException",
            "TelemetrySenderSendWithCloudEventIdEmpty_ThrowsException",
            "TelemetrySenderSendWithCloudEventSourceNonUri_ThrowsException",
            "TelemetrySenderSendWithCloudEventSpecVersionEmpty_ThrowsException",
            "TelemetrySenderSendWithCloudEventSubjectEmpty_ThrowsException",
            "TelemetrySenderSendWithCloudEventTypeEmpty_ThrowsException",
        };

        private static IDeserializer yamlDeserializer;
        private static AsyncAtomicInt TestCaseIndex = new(0);

        static TelemetrySenderTester()
        {
            yamlDeserializer = new DeserializerBuilder()
                .WithNamingConvention(HyphenatedNamingConvention.Instance)
                .WithEnumNamingConvention(HyphenatedNamingConvention.Instance)
                .WithTypeDiscriminatingNodeDeserializer(options =>
                {
                    options.AddKeyValueTypeDiscriminator<TestCaseAction>("action",
                        ("send telemetry", typeof(TestCaseActionSendTelemetry)),
                        ("await send", typeof(TestCaseActionAwaitSend)),
                        ("await publish", typeof(TestCaseActionAwaitPublish)),
                        ("disconnect", typeof(TestCaseActionDisconnect)));
                })
                .Build();

            if (File.Exists(defaultsFilePath))
            {
                DefaultTestCase defaultTestCase = Toml.ToModel<DefaultTestCase>(File.ReadAllText(defaultsFilePath), defaultsFilePath, new TomlModelOptions { ConvertPropertyName = CaseConverter.PascalToKebabCase });

                TestCaseSerializer.DefaultOutContentType = defaultTestCase.Prologue.Sender.Serializer.OutContentType;
                TestCaseSerializer.DefaultIndicateCharacterData = defaultTestCase.Prologue.Sender.Serializer.IndicateCharacterData;

                TestCaseSender.DefaultTelemetryName = defaultTestCase.Prologue.Sender.TelemetryName;
                TestCaseSender.DefaultTelemetryTopic = defaultTestCase.Prologue.Sender.TelemetryTopic;
                TestCaseSender.DefaultTopicNamespace = defaultTestCase.Prologue.Sender.TopicNamespace;

                TestCaseActionSendTelemetry.DefaultTelemetryName = defaultTestCase.Actions.SendTelemetry.TelemetryName;
                TestCaseActionSendTelemetry.DefaultTelemetryValue = defaultTestCase.Actions.SendTelemetry.TelemetryValue;
                TestCaseActionSendTelemetry.DefaultTimeout = defaultTestCase.Actions.SendTelemetry.Timeout;
                TestCaseActionSendTelemetry.DefaultQos = defaultTestCase.Actions.SendTelemetry.Qos;
            }
        }

        public static IEnumerable<object[]> GetAllTelemetrySenderCases()
        {
            foreach (string testCasePath in Directory.GetFiles(senderCasesPath, @"*.yaml"))
            {
                string testCaseName = Path.GetFileNameWithoutExtension(testCasePath);
                using (StreamReader streamReader = File.OpenText($"{senderCasesPath}/{testCaseName}.yaml"))
                {
                    TestCase testCase = yamlDeserializer.Deserialize<TestCase>(new Parser(streamReader));
                    if (!problematicTestCases.Contains(testCaseName) &&
                        !testCase.Requires.Contains(TestFeatureKind.Unobtanium))
                    {
                        yield return new object[] { testCaseName };
                    }
                }
            }
        }

        public static IEnumerable<object[]> GetRestrictedTelemetrySenderCases()
        {
            foreach (string testCasePath in Directory.GetFiles(senderCasesPath, @"*.yaml"))
            {
                string testCaseName = Path.GetFileNameWithoutExtension(testCasePath);
                using (StreamReader streamReader = File.OpenText($"{senderCasesPath}/{testCaseName}.yaml"))
                {
                    TestCase testCase = yamlDeserializer.Deserialize<TestCase>(new Parser(streamReader));
                    if (!problematicTestCases.Contains(testCaseName) &&
                        !testCase.Requires.Contains(TestFeatureKind.Unobtanium) &&
                        !testCase.Requires.Contains(TestFeatureKind.AckOrdering) &&
                        !testCase.Requires.Contains(TestFeatureKind.Reconnection))
                    {
                        yield return new object[] { testCaseName };
                    }
                }
            }
        }

        [Theory]
        [MemberData(nameof(GetAllTelemetrySenderCases))]
        public Task TestTelemetrySenderWithSessionClient(string testCaseName)
        {
            return TestTelemetrySenderProtocol(testCaseName, includeSessionClient: true);
        }

        [Theory]
        [MemberData(nameof(GetRestrictedTelemetrySenderCases))]
        public Task TestTelemetrySenderStandalone(string testCaseName)
        {
            return TestTelemetrySenderProtocol(testCaseName, includeSessionClient: false);
        }

        private async Task TestTelemetrySenderProtocol(string testCaseName, bool includeSessionClient)
        {
            int testCaseIndex = await TestCaseIndex.Increment().ConfigureAwait(false);

            TestCase testCase;
            using (StreamReader streamReader = File.OpenText($"{senderCasesPath}/{testCaseName}.yaml"))
            {
                testCase = yamlDeserializer.Deserialize<TestCase>(new Parser(streamReader));
            }

            Dictionary<string, TestTelemetrySender> telemetrySenders = new();

            string clientIdPrefix = includeSessionClient ? "Session" : "Standalone";
            string mqttClientId = testCase.Prologue?.MqttConfig?.ClientId ?? $"{clientIdPrefix}SenderTestClient{testCaseIndex}";
            StubMqttClient stubMqttClient = new StubMqttClient(mqttClientId);
            await using CompositeMqttClient compositeMqttClient = new CompositeMqttClient(stubMqttClient, includeSessionClient, mqttClientId);

            await compositeMqttClient.ConnectAsync().WaitAsync(TestTimeout);

            if (testCase.Prologue?.PushAcks != null)
            {
                foreach (TestAckKind ackKind in testCase.Prologue.PushAcks.Publish)
                {
                    stubMqttClient.EnqueuePubAck(ackKind);
                }

                foreach (TestAckKind ackKind in testCase.Prologue.PushAcks.Subscribe)
                {
                    stubMqttClient.EnqueueSubAck(ackKind);
                }

                foreach (TestAckKind ackKind in testCase.Prologue.PushAcks.Unsubscribe)
                {
                    stubMqttClient.EnqueueUnsubAck(ackKind);
                }
            }

            foreach (TestCaseSender testCaseSender in testCase.Prologue?.Senders ?? new List<TestCaseSender>())
            {
                bool isLast = ReferenceEquals(testCaseSender, testCase.Prologue?.Senders.Last());
                TestTelemetrySender? telemetrySender = await GetTelemetrySender(compositeMqttClient, testCaseSender, isLast ? testCase.Prologue?.Catch : null);
                if (telemetrySender == null)
                {
                    return;
                }

                telemetrySenders[testCaseSender.TelemetryName!] = telemetrySender;
            }

            AsyncQueue<Task> sendTasks = new();

            foreach (TestCaseAction action in testCase.Actions)
            {
                switch (action)
                {
                    case TestCaseActionSendTelemetry actionSendTelemetry:
                        SendTelemetryAsync(actionSendTelemetry, telemetrySenders, sendTasks);
                        break;
                    case TestCaseActionAwaitSend actionAwaitSend:
                        await AwaitSendAsync(actionAwaitSend, sendTasks).ConfigureAwait(false);
                        break;
                    case TestCaseActionAwaitPublish:
                        await AwaitPublishAsync(stubMqttClient).ConfigureAwait(false);
                        break;
                    case TestCaseActionDisconnect:
                        await DisconnectAsync(stubMqttClient).ConfigureAwait(false);
                        break;
                }
            }

            while (sendTasks.TryDequeue(out Task? sendTask))
            {
                try
                {
                    await sendTask.WaitAsync(TestTimeout).ConfigureAwait(false);
                }
                catch (AkriMqttException exception)
                {
                    Assert.Fail($"Unexpected exception thrown when awaiting TelemetrySender.SendTelemetryAsync(): {exception.Message}");
                }
            }

            if (testCase.Epilogue != null)
            {
                foreach (string topic in testCase.Epilogue.SubscribedTopics)
                {
                    Assert.True(stubMqttClient.HasSubscribed(topic));
                }

                if (testCase.Epilogue.PublicationCount != null)
                {
                    int publicationCount = await stubMqttClient.GetPublicationCount().ConfigureAwait(false);
                    Assert.Equal(testCase.Epilogue.PublicationCount, publicationCount);
                }

                int sequenceIndex = 0;
                foreach (TestCasePublishedMessage publishedMessage in testCase.Epilogue.PublishedMessages)
                {
                    CheckPublishedMessage(sequenceIndex, publishedMessage, stubMqttClient);
                    sequenceIndex++;
                }

                if (testCase.Epilogue.AcknowledgementCount != null)
                {
                    int acknowledgementCount = await stubMqttClient.GetAcknowledgementCount().ConfigureAwait(false);
                    Assert.Equal(testCase.Epilogue.AcknowledgementCount, acknowledgementCount);
                }
            }
        }

        private async Task<TestTelemetrySender?> GetTelemetrySender(IMqttPubSubClient mqttClient, TestCaseSender testCaseSender, TestCaseCatch? testCaseCatch)
        {
            try
            {
                TestSerializer testSerializer = new TestSerializer(testCaseSender.Serializer);

                TestTelemetrySender telemetrySender = new TestTelemetrySender(new ApplicationContext(), mqttClient, testCaseSender.TelemetryName!, testSerializer)
                {
                    TopicPattern = testCaseSender.TelemetryTopic!,
                    TopicNamespace = testCaseSender.TopicNamespace,
                };

                if (testCaseSender.TopicTokenMap != null)
                {
                    foreach (KeyValuePair<string, string> kvp in testCaseSender.TopicTokenMap)
                    {
                        telemetrySender.TopicTokenMap![kvp.Key] = kvp.Value;
                    }
                }

                if (testCaseCatch != null)
                {
                    // TelemetrySender has no Start method, so if an exception is expected, Send may be needed to trigger it.
                    try
                    {
                        await telemetrySender.SendTelemetryAsync(TestCaseActionSendTelemetry.DefaultTelemetryValue!).WaitAsync(TestTimeout);
                    }
                    catch (AkriMqttException exception)
                    {
                        if (exception.Kind != AkriMqttErrorKind.Cancellation)
                        {
                            AkriMqttExceptionChecker.CheckException(testCaseCatch, exception);
                            return null;
                        }
                    }

                    Assert.Fail($"Expected {testCaseCatch.ErrorKind} exception, but no exception thrown when initializing TelemetrySender");
                }

                return telemetrySender;
            }
            catch (AkriMqttException exception)
            {
                if (testCaseCatch == null)
                {
                    Assert.Fail($"Unexpected exception thrown initializing TelemetrySender: {exception.Message}");
                }

                AkriMqttExceptionChecker.CheckException(testCaseCatch, exception);
                return null;
            }
        }

        private void SendTelemetryAsync(TestCaseActionSendTelemetry actionSendTelemetry, Dictionary<string, TestTelemetrySender> telemetrySenders, AsyncQueue<Task> sendTasks)
        {
            OutgoingTelemetryMetadata metadata = new OutgoingTelemetryMetadata();

            if (actionSendTelemetry.Metadata != null)
            {
                foreach (KeyValuePair<string, string> kvp in actionSendTelemetry.Metadata)
                {
                    metadata.UserData[kvp.Key] = kvp.Value;
                }
            }

            if (actionSendTelemetry.CloudEvent != null)
            {
                Uri sourceUri = new Uri(actionSendTelemetry.CloudEvent.Source!, UriKind.RelativeOrAbsolute);

                if (actionSendTelemetry.CloudEvent.Type != null && actionSendTelemetry.CloudEvent.SpecVersion != null)
                {
                    metadata.CloudEvent = new CloudEvent(sourceUri, actionSendTelemetry.CloudEvent.Type, actionSendTelemetry.CloudEvent.SpecVersion);
                }
                else if (actionSendTelemetry.CloudEvent.Type != null)
                {
                    metadata.CloudEvent = new CloudEvent(sourceUri, type: actionSendTelemetry.CloudEvent.Type);
                }
                else if (actionSendTelemetry.CloudEvent.SpecVersion != null)
                {
                    metadata.CloudEvent = new CloudEvent(sourceUri, specversion: actionSendTelemetry.CloudEvent.SpecVersion);
                }
                else
                {
                    metadata.CloudEvent = new CloudEvent(sourceUri);
                }

                if (actionSendTelemetry.CloudEvent.Id != null)
                {
                    metadata.CloudEvent.Id = actionSendTelemetry.CloudEvent.Id;
                }

                if (actionSendTelemetry.CloudEvent.Time is string timeString)
                {
                    if (DateTime.TryParse(timeString, null, DateTimeStyles.RoundtripKind, out DateTime time))
                    {
                        metadata.CloudEvent.Time = time;
                    }
                    else
                    {
                        sendTasks.Enqueue(Task.FromException(AkriMqttException.GetArgumentInvalidException(null, "CloudEvent", null, null)));
                        return;
                    }
                }

                if (actionSendTelemetry.CloudEvent.Subject is string subject)
                {
                    metadata.CloudEvent.Subject = subject;
                }

                if (actionSendTelemetry.CloudEvent.DataSchema is string dataSchema)
                {
                    metadata.CloudEvent.DataSchema = dataSchema;
                }
            }

            MqttQualityOfServiceLevel qos = actionSendTelemetry.Qos != null ? (MqttQualityOfServiceLevel)actionSendTelemetry.Qos : MqttQualityOfServiceLevel.AtLeastOnce;
            sendTasks.Enqueue(telemetrySenders[actionSendTelemetry.TelemetryName!].SendTelemetryAsync(actionSendTelemetry.TelemetryValue!, metadata, qos, actionSendTelemetry.Timeout?.ToTimeSpan()));
        }

        private async Task AwaitSendAsync(TestCaseActionAwaitSend actionAwaitSend, AsyncQueue<Task> sendTasks)
        {
            try
            {
                Task sendTask = await sendTasks.DequeueAsync().WaitAsync(TestTimeout).ConfigureAwait(false);
                await sendTask.WaitAsync(TestTimeout).ConfigureAwait(false);

                if (actionAwaitSend.Catch != null)
                {
                    Assert.Fail($"Expected {actionAwaitSend.Catch.ErrorKind} exception, but no exception thrown when awaiting TelemetrySender.SendTelemetryAsync()");
                }
            }
            catch (AkriMqttException exception)
            {
                if (actionAwaitSend.Catch == null)
                {
                    Assert.Fail($"Unexpected exception thrown when awaiting TelemetrySender.SendTelemetryAsync(): {exception.Message}");
                }

                AkriMqttExceptionChecker.CheckException(actionAwaitSend.Catch, exception);
            }
        }

        private async Task AwaitPublishAsync(StubMqttClient stubMqttClient)
        {
            await stubMqttClient.AwaitPublishAsync().WaitAsync(TestTimeout).ConfigureAwait(false);
        }

        private Task DisconnectAsync(StubMqttClient stubMqttClient)
        {
            return stubMqttClient.DisconnectAsync(new MQTTnet.Client.MqttClientDisconnectOptions());
        }

        private void CheckPublishedMessage(int sequenceIndex, TestCasePublishedMessage publishedMessage, StubMqttClient stubMqttClient)
        {
            MQTTnet.MqttApplicationMessage? appMsg = stubMqttClient.GetPublishedMessage(sequenceIndex);
            Assert.NotNull(appMsg);

            if (publishedMessage.Topic != null)
            {
                Assert.Equal(publishedMessage.Topic, appMsg.Topic);
            }

            if (publishedMessage.Payload == null)
            {
                Assert.Null(appMsg.PayloadSegment.Array);
            }
            else if (publishedMessage.Payload is string payload)
            {
                Assert.NotNull(appMsg.PayloadSegment.Array);
                Assert.Equal(payload, Encoding.UTF8.GetString(appMsg.PayloadSegment.Array));
            }

            if (publishedMessage.ContentType != null)
            {
                Assert.Equal(publishedMessage.ContentType, appMsg.ContentType);
            }

            if (publishedMessage.FormatIndicator != null)
            {
                Assert.Equal(publishedMessage.FormatIndicator, (int)appMsg.PayloadFormatIndicator);
            }

            foreach (KeyValuePair<string, string?> kvp in publishedMessage.Metadata)
            {
                if (kvp.Value != null)
                {
                    Assert.True(MqttNetConverter.ToGeneric(appMsg.UserProperties).TryGetProperty(kvp.Key, out string? value), $"header {kvp.Key} not present");
                    Assert.Equal(kvp.Value, value);
                }
                else
                {
                    Assert.False(MqttNetConverter.ToGeneric(appMsg.UserProperties).TryGetProperty(kvp.Key, out string? value), $"header {kvp.Key} unexpectedly present with value '{value}'");
                }
            }

            if (publishedMessage.SourceId != null)
            {
                Assert.True(MqttNetConverter.ToGeneric(appMsg.UserProperties).TryGetProperty(AkriSystemProperties.SourceId, out string? sourceId));
                Assert.Equal(publishedMessage.SourceId, sourceId);
            }

            if (publishedMessage.Expiry != null)
            {
                Assert.Equal((uint)publishedMessage.Expiry, appMsg.MessageExpiryInterval);
            }
        }
    }
}
